pkg/httpc/client.go

```diff
@@ -102,22 +98,44 @@ func (c *Client) Send(
 		return nil, e
 	}
 
-	if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
+	var data []byte
+	if !c.isRawBody {
+		data, err = ioutil.ReadAll(resp.Body)
 		defer resp.Body.Close()
-		data, _ := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			e := errType.Wrap(err, "Failed to read %s API response", errOriginComponent)
+			log.Warn("SendRequest failed", zap.String("uri", uri), zap.Error(err))
+			return nil, e
+		}
+	}
+
+	if !(resp.StatusCode >= 200 && resp.StatusCode < 300) {
 		e := errType.New("Request failed with status code %d from %s API: %s", resp.StatusCode, errOriginComponent, string(data))
 		log.Warn("SendRequest failed", zap.String("uri", uri), zap.Error(err))
 		return nil, e
 	}
 
-	return &Response{resp}, nil
+	return &Response{RawResponse: resp, body: data}, nil
 }
 
 type Response struct {
-	*http.Response
+	RawResponse *http.Response
+
+	body []byte
 }
 
-func (r *Response) Body() ([]byte, error) {
-	defer r.Response.Body.Close()
-	return ioutil.ReadAll(r.Response.Body)
+func (r *Response) Body() []byte {
+	if r.RawResponse == nil {
+		return []byte{}
+	}
+	return r.body
+}
+
+// RawBody method exposes the HTTP raw response body. Use this method in-conjunction with `WithRawBody()`
+// function otherwise you get an error as `read err: http: read on closed response body`.
+func (r *Response) RawBody() io.ReadCloser {
```

Comment:
Seems that the change of this file's API doesn't reduce the complexity, but make the calling more complex:

Previous design:

| Action                  | Function Call before Send | Function Call for Response |
|-------------------------|---------------------------|----------------------------|
| Access Raw Body         | No need.                  | #.Body                     |
| Access Body as Bytes    | No need.                  | #.Body()                   |
| Access Response headers | No need.                  | #.Headers                  |

New design:

| Action                  | Function Call before Send | Function Call for Response |
|-------------------------|---------------------------|----------------------------|
| Access Raw Body         | WithRawBody()             | #.RawBody()                |
| Access Body as Bytes    | No need.                  | #.Body()                   |
| Access Response headers | No need.                  | #.Response.Headers         |

There is also possibilities that people can now call `WithRawBody()` + `Body()` that will just lead to empty data, or `RawBody()` directly that will lead to `nil`. None of these can be ensured by the compiler: more possibilities to make mistakes.

What do you want to achieve by this change?

Reply:
Yes, you're right.
I wanted to unify the previous interfaces `Send` and `SendRequest`. So I learned the deisgn from go-resty and introduced a risk at the same time, which is not worth it.

Maybe renaming `SendRequest` to `SendAndReadBody` would work.

Reply:
Yes. The most uncomfortable part I fell when using go-resty is that I need to **pair** the raw request and the raw response otherwise it just won't work.